lucille-core/src/main/java/com/kmwllc/lucille/core/Document.java

```diff
@@ -570,4 +574,150 @@ static Document createFromJson(String json, UnaryOperator<String> idUpdater) thr
+   * This also works for nested values that contain a list, such as "a.b.2.c" where 'b' is an ArrayNode. The indices are 0 based.
+   *
+   * @param name the nested field path to set the JsonNode at
+   * @param value the JsonNode to set at the nested path
+   */
+  default void setNestedJson(String name, JsonNode value) {
+    setNestedJson(name.split("\\."), value);
+  }
+
+  /**
+   * Sets a nested JsonNode at the given path segments where each part is treated as a level of nesting.
+   * This works for nested values that contain a list, such as ["a","b","2","c"] where 'b' is an ArrayNode. The indices are 0 based.
+   *
+   * @param pathSegments the nested field path segments to set the JsonNode at
+   * @param value the JsonNode to set at the nested path
+   */
+  default void setNestedJson(String[] pathSegments, JsonNode value) {
+    if (pathSegments.length == 0) {
+      return;
+    }
+    if (pathSegments.length == 1) {
+      setField(pathSegments[0], value);
+      return;
+    }
+
+    // if necessary, create a default fresh node where the type is based off of the child field, either ArrayNode or ObjectNode
+    JsonNode node = has(pathSegments[0]) ? getJson(pathSegments[0]) : createNewNode(pathSegments[1]);
+
+    JsonNode currentNode = node;
+    for (int i = 1; i < pathSegments.length-1; i++) {
+      String segment = pathSegments[i];
+
+      if (hasFieldSegment(currentNode, segment)) {
+        currentNode = getFieldSegment(currentNode,segment);
+      } else {
+        JsonNode childNode = createNewNode(pathSegments[i+1]);
+        setFieldSegment(currentNode, segment, childNode);
+        currentNode = childNode;
+      }
+    }
+
+    // set last node
+    setFieldSegment(currentNode, pathSegments[pathSegments.length-1], value);
+
+    setField(pathSegments[0], node);
+  }
+
+  // if the given segment is an integer, it will create a new ArrayNode, otherwise an ObjectNode
+  private static JsonNode createNewNode(String segment) {
+    try { // deal with possible array nodes
+      Integer.parseInt(segment);
+      return JsonDocument.MAPPER.createArrayNode();
+    } catch (NumberFormatException nfe) {
+      // not an integer, so leave as ObjectNode
+      return JsonDocument.MAPPER.createObjectNode();
+    }
+  }
+
+  // sets the given value on the given field segment with either an ObjectNode or ArrayNode
+  private static void setFieldSegment(JsonNode node, String segment, JsonNode value) {
+    if (node.isArray()) {
+      int index = Integer.parseInt(segment);
+      // check if index is within the existing array and overwrite if so
+      if (index >= 0 && index < node.size()) {
+        ((ArrayNode) node).set(index, value);
+      } else if (index == node.size()) { // add to end of array (including creating as array if index==0)
+        ((ArrayNode) node).add(value);
+      } else {
+        throw new IllegalArgumentException("Cannot set index " + index + " on array of size " + node.size());
```

Comment:
thinking about this more... I would change this to ArrayIndexOutOfBoundsException because it's more descriptive. It's also an unchecked exception so you don't need to declare it.

I agree with what you did in AddRandomNestedField (catch & rethrow with more info) and CopyFields (only log an error)